File: Neetcode/Add2Nums.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        h1 = l1
        h2 = l2
        carry = 0
        head = ListNode(-1)
        start = head

        while h1 is not None and h2 is not None:
            if carry:
                add = h1.val + h2.val + 1
            else:
                add = h1.val + h2.val
            carry = add//10
            head.next = ListNode(add%10)
            head = head.next
            h1 = h1.next
            h2 = h2.next

        if h1 is not None:
            head.next = h1
            head = head.next
        elif h2 is not None:
            head.next = h2
            head = head.next
        else:
            if carry:
                head.next = ListNode(1)
            return start.next


        while head is not None and carry:
            add = head.val + carry
            carry = add//10
            head.val = add%10
            temp = head
            head = head.next
        
        if carry:
            temp.next = ListNode(1)
        
        return start.next

File: Neetcode/Add2Nums.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int ans = 0;
        int carry = 0;
        ListNode* dummy = new ListNode(-1);
        ListNode* d = dummy;
        while (l1 != nullptr && l2 != nullptr){
            ans = l1->val + l2->val;
            if (carry){
                ans += 1;
                carry = 0;
            }
            carry = ans/10;
            d->next = new ListNode(ans%10);
            d = d->next;
            l1 = l1->next;
            l2 = l2->next;
        }
        while (carry){
            if (l1 != nullptr){
                ans = l1->val + carry;
                carry = 0;
                l1 = l1->next;
            }
            else if (l2 != nullptr){
                ans = l2->val + carry;
                carry = 0;
                l2 = l2->next;
            }
            else{
                if (carry){
                    ans = 1;
                }
            }
            carry = ans/10;
            d->next = new ListNode(ans%10);
            d = d->next;
        }
        while (l1 != nullptr){
            ans = l1->val;
            d->next = new ListNode(ans);
            d = d->next;
            l1 = l1->next;
        }
        while (l2 != nullptr){
            ans = l2->val;
            d->next = new ListNode(ans);
            d = d->next;
            l2 = l2->next;
        }
        return dummy->next;
    }
};
